xptifw/src/xpti_trace_framework.cpp

```diff
@@ -532,107 +446,61 @@ class Tracepoints {
-        // it is on the stack
-        if (InstanceNo)
-          *InstanceNo = EvLoc->second.instance_id;
-        return &(EvLoc->second);
-      } else
-        return nullptr; // we have a problem!
+      EvLoc->second.instance_id++;
+      // Guarantees that the returned instance ID will be accurate as
+      // it is on the stack
+      if (InstanceNo)
+        *InstanceNo = EvLoc->second.instance_id;
+      return &(EvLoc->second);
     } else {
 #ifdef XPTI_STATISTICS
       MInsertions++;
 #endif
-      // Create a new unique ID
-      //
-      UId = MUId++;
-      // And add it as a pair
-      //
-      MPayloadLUT.add(HashValue, UId);
-      // The API allows you to query a Universal ID from the kernel address; so
-      // build the necessary data structures for this.
-      if (TempPayload.flags & (uint64_t)payload_flag_t::HashAvailable) {
-#ifndef XPTI_USE_TBB
-        std::lock_guard<std::mutex> Lock(MCodePtrMutex);
-#endif
-        MCodePtrLUT[reinterpret_cast<uint64_t>(TempPayload.code_ptr_va)] = UId;
-      }
       // We also want to query the payload by universal ID that has been
       // generated
-#ifndef XPTI_USE_TBB
-      std::lock_guard<std::mutex> Lock(MMutex);
-#endif
-      MPayloads[UId] = TempPayload; // when it uses tbb, should be thread-safe
-      {
-        xpti::trace_event_data_t *Event = &MEvents[UId];
-        // We are seeing this unique ID for the first time, so we will
-        // initialize the event structure with defaults and set the unique_id to
-        // the newly generated unique id (uid)
-        Event->unique_id = UId;
-        Event->unused = 0;
-        Event->reserved.payload = &MPayloads[UId];
-        Event->data_id = Event->source_id = Event->target_id = 0;
-        Event->instance_id = 1;
-        Event->global_user_data = nullptr;
-        Event->event_type = (uint16_t)xpti::trace_event_type_t::unknown_event;
-        Event->activity_type =
-            (uint16_t)xpti::trace_activity_type_t::unknown_activity;
-        *InstanceNo = Event->instance_id;
-        return Event;
-      }
+      auto &CurrentPayload = MPayloads[HashValue];
+      CurrentPayload = TempPayload; // when it uses tbb, should be thread-safe
+      xpti::trace_event_data_t *Event = &MEvents[HashValue];
+      // We are seeing this unique ID for the first time, so we will
+      // initialize the event structure with defaults and set the unique_id to
+      // the newly generated unique id (uid)
+      Event->unique_id = HashValue;
+      Event->unused = 0;
+      Event->reserved.payload = &CurrentPayload;
+      Event->data_id = Event->source_id = Event->target_id = 0;
+      Event->instance_id = 1;
+      Event->global_user_data = nullptr;
+      Event->event_type = (uint16_t)xpti::trace_event_type_t::unknown_event;
+      Event->activity_type =
+          (uint16_t)xpti::trace_activity_type_t::unknown_activity;
+      *InstanceNo = Event->instance_id;
+      return Event;
     }
-    return nullptr;
```

Comment:
It looks like all cases are handled, but I'm still not sure if all compilers can process this code without firing a warning about non-void function without return statement.

Reply:
If/else return values in all cases and the compilers did not complain. I prefer this as any changes in the future that accidentally delete a return line will cause the compiler to complain.